File: frontend/src/app/services/statistics.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable, map } from 'rxjs';

@Injectable({ providedIn: 'root' })
export class StatisticsService {
  private baseUrl = 'http://localhost:8080/stats';

  constructor(private http: HttpClient) {}

  // ===== KPI =====
  getTotalOrders(year?: number): Observable<number> {
    let params = new HttpParams();
    if (year) params = params.set('year', year);
    return this.http.get<any>(`${this.baseUrl}/total-orders`, { params }).pipe(map(v => Number(v.value ?? 0)));
  }

  getTotalCustomers(): Observable<number> {
    return this.http.get<any>(`${this.baseUrl}/total-customers`).pipe(map(v => Number(v.value ?? 0)));
  }

  getTotalRevenue(year?: number): Observable<number> {
    let params = new HttpParams();
    if (year) params = params.set('year', year);
    return this.http.get<any>(`${this.baseUrl}/total-revenue`, { params }).pipe(map(v => Number(v.value ?? 0)));
  }

  getItemsPerOrder(): Observable<number> {
    return this.http.get<any>(`${this.baseUrl}/items-per-order`).pipe(map(v => Number(v.value ?? 0)));
  }

  // ===== TIME =====
  getOrdersPerMonth(year?: number): Observable<any[]> {
    let params = new HttpParams();
    if (year) params = params.set('year', year);
    return this.http.get<any[]>(`${this.baseUrl}/orders-per-month`, { params });
  }

  getRevenueByMonth(year?: number): Observable<any[]> {
    let params = new HttpParams();
    if (year) params = params.set('year', year);
    return this.http.get<any[]>(`${this.baseUrl}/revenue-by-month`, { params });
  }

  // ===== PRODUCTS =====
  getTopProducts(limit = 10): Observable<any[]> {
    const params = new HttpParams().set('limit', limit);
    return this.http.get<any[]>(`${this.baseUrl}/top-products`, { params });
  }

  getTopProductsMonth(month: string, limit = 10): Observable<any[]> {
    const params = new HttpParams().set('month', month).set('limit', limit);
    return this.http.get<any[]>(`${this.baseUrl}/top-products/month`, { params });
  }

  getTopProductsYear(year: number, limit = 10): Observable<any[]> {
    const params = new HttpParams().set('year', year).set('limit', limit);
    return this.http.get<any[]>(`${this.baseUrl}/top-products/year`, { params });
  }

  getTopProductsByRevenue(limit = 10): Observable<any[]> {
    const params = new HttpParams().set('limit', limit);
    return this.http.get<any[]>(`${this.baseUrl}/top-products/revenue`, { params });
  }

  // ===== CATEGORY =====
  getCategoryShare(year?: number): Observable<any[]> {
    let params = new HttpParams();
    if (year) params = params.set('year', year);
    return this.http.get<any[]>(`${this.baseUrl}/category/share`, { params });
  }

  getCategoryMargin(year?: number): Observable<any[]> {
    let params = new HttpParams();
    if (year) params = params.set('year', year);
    return this.http.get<any[]>(`${this.baseUrl}/category/margin`, { params });
  }
}

File: backend/src/main/java/com/electronics/backend/services/StatisticsService.java
package com.electronics.backend.services;

import com.electronics.backend.repository.StatisticsRepository;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Map;

@Service
public class StatisticsService {

    private final StatisticsRepository repo;

    public StatisticsService(StatisticsRepository repo) {
        this.repo = repo;
    }

    // ===== KPI =====
    public long getTotalOrders() { return repo.getTotalOrders(); }
    public long getTotalOrdersByYear(int year) { return repo.getTotalOrdersByYear(year); }

    public long getTotalCustomers() { return repo.getTotalCustomers(); }

    public double getTotalRevenue() { return repo.getTotalRevenue(); }
    public double getTotalRevenueByYear(int year) { return repo.getTotalRevenueByYear(year); }

    public double getItemsPerOrder() { return repo.getItemsPerOrder(); }

    // ===== TIME =====
    public List<Map<String, Object>> getRevenueByMonth(Integer year) { return repo.getRevenueByMonth(year); }
    public List<Map<String, Object>> getOrdersPerMonth(Integer year) { return repo.getOrderCountByMonth(year); }

    // ===== PRODUCTS =====
    public List<Map<String, Object>> getTopProductsOverall(int limit) { return repo.getTopProductsOverall(limit); }
    public List<Map<String, Object>> getTopProductsByMonth(String month, int limit) { return repo.getTopProductsByMonth(month, limit); }
    public List<Map<String, Object>> getTopProductsByYear(int year, int limit) { return repo.getTopProductsByYear(year, limit); }
    public List<Map<String, Object>> getTopProductsByRevenue(int limit) { return repo.getTopProductsByRevenue(limit); }

    // ===== CLIENTS =====
    public List<Map<String, Object>> getTopClients(int limit) { return repo.getTopClients(limit); }

    // ===== CATEGORY =====
    public List<Map<String, Object>> getRevenueByCategory(Integer year) { return repo.getRevenueByCategory(year); }
    public List<Map<String, Object>> getMarginByCategory(Integer year) { return repo.getMarginByCategory(year); }
    public List<Map<String, Object>> getCategoryShare(Integer year) { return repo.getCategoryShare(year); }
}
